File: Data Structures & Algorithms/maximum-product-subarray/submission-4.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n, pre, suf, ans = len(nums), 1, 1, float('-inf')

        for i in range(n):
            pre *= nums[i]
            suf *= nums[n - i - 1]

            ans = max(ans, pre, suf)
            if pre == 0:
                pre = 1
            if suf == 0:
                suf = 1
        
        return ans

File: Data Structures & Algorithms/jump-game-ii/submission-4.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    jump(nums) {
        let ans = 0, far = 0, end = 0;
        for(let i = 0; i < nums.length - 1; ++i) {
            far = Math.max(far, nums[i] + i);
            if(i == end) {
                ans++;
                end = far;
            }
        }
        return ans;
    }
}
